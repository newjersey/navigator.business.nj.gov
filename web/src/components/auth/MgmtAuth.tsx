import * as apiClient from "@/lib/api-client/apiClient";
import { TextField } from "@mui/material";
import { ChangeEvent, KeyboardEvent, ReactElement, useState } from "react";

interface Props {
  password: string;
  setPassword: (value: string) => void;
  setIsAuthed: (value: boolean) => void;
}

export const MgmtAuth = (props: Props): ReactElement => {
  const { password } = props;
  const [hasError, setHasError] = useState<boolean>(false);

  const handleKeyPress = (event: KeyboardEvent<HTMLDivElement>): void => {
    if (event.code === "Enter" || event.key === "Enter") {
      onSubmit();
    }
  };

  const handlePasswordInput = (event: ChangeEvent<HTMLInputElement>): void => {
    setHasError(false);
    props.setPassword(event.target.value);
  };

  const onSubmit = (): void => {
    setHasError(false);
    apiClient
      .post("/mgmt/auth", { password }, false)
      .then(() => {
        return props.setIsAuthed(true);
      })
      .catch(() => {
        setHasError(true);
      });
  };

  return (
    <>
      <h1>Enter admin password:</h1>
      <label htmlFor="password">Password</label>
      <TextField
        name="password"
        variant="outlined"
        type="password"
        value={password}
        error={hasError}
        helperText={hasError ? "Authentication failed" : ""}
        onChange={handlePasswordInput}
        onKeyPress={handleKeyPress}
        inputProps={{
          id: "password",
          "data-testid": "mgmt-password-field",
        }}
      />
      <button
        onClick={onSubmit}
        className="usa-button margin-top-2"
        data-testid="mgmt-submit-bttn"
      >
        Submit
      </button>
    </>
  );
};
