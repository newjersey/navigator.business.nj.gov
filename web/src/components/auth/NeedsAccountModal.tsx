import { Content } from "@/components/Content";
import { ModalZeroButton } from "@/components/ModalZeroButton";
import { PrimaryButton } from "@/components/njwds-extended/PrimaryButton";
import { SecondaryButton } from "@/components/njwds-extended/SecondaryButton";
import { NeedsAccountContext } from "@/contexts/needsAccountContext";
import { IsAuthenticated } from "@/lib/auth/AuthContext";
import { triggerSignIn } from "@/lib/auth/sessionHelper";
import { useConfig } from "@/lib/data-hooks/useConfig";
import { useUserData } from "@/lib/data-hooks/useUserData";
import { ROUTES } from "@/lib/domain-logic/routes";
import analytics from "@/lib/utils/analytics";
import { useMountEffectWhenDefined } from "@/lib/utils/helpers";
import { useRouter } from "next/compat/router";
import { ReactElement, useContext } from "react";

export const NeedsAccountModal = (): ReactElement => {
  const { business } = useUserData();
  const router = useRouter();
  const {
    isAuthenticated,
    showNeedsAccountModal,
    showContinueWithoutSaving,
    setShowNeedsAccountModal,
    setUserWantsToContinueWithoutSaving,
  } = useContext(NeedsAccountContext);
  const { Config } = useConfig();
  const loginPageEnabled = process.env.FEATURE_LOGIN_PAGE === "true";

  useMountEffectWhenDefined(() => {
    if (isAuthenticated === IsAuthenticated.TRUE) {
      setShowNeedsAccountModal(false);
    }
  }, isAuthenticated);

  if (isAuthenticated === IsAuthenticated.TRUE) {
    return <></>;
  }

  const linkToAccountSetup = (): void => {
    if (business?.preferences.returnToLink === `${ROUTES.dashboard}`) {
      analytics.event.myNJ_prompt_modal_complete_button.click.go_to_NavigatorAccount_setup();
    } else {
      analytics.event.guest_modal.click.go_to_NavigatorAccount_setup();
    }
    setShowNeedsAccountModal(false);
    router && router.push(ROUTES.accountSetup);
  };

  return (
    <>
      <ModalZeroButton
        isOpen={showNeedsAccountModal}
        close={(): void => setShowNeedsAccountModal(false)}
        title={Config.selfRegistration.needsAccountModalTitle}
      >
        <div data-testid="self-reg-modal">
          <Content>{Config.selfRegistration.needsAccountModalBody}</Content>
          <div className="margin-top-3">
            <PrimaryButton isColor="primary" isFullWidthOnDesktop onClick={linkToAccountSetup}>
              {Config.selfRegistration.needsAccountModalButtonText}
            </PrimaryButton>
            {showContinueWithoutSaving && (
              <SecondaryButton
                isColor="primary"
                className={"margin-top-05"}
                isFullWidthOnDesktop
                onClick={() => {
                  setUserWantsToContinueWithoutSaving(true);
                  setShowNeedsAccountModal(false);
                }}
              >
                {Config.selfRegistration.continueWithoutSaving}
              </SecondaryButton>
            )}
          </div>

          <hr className="margin-y-3 margin-x-neg-4" />

          <div className="flex flex-column flex-align-center">
            <Content
              onClick={(): void => {
                analytics.event.guest_modal.click.go_to_myNJ_login();
                if (loginPageEnabled) {
                  router && router.push(ROUTES.login);
                } else {
                  triggerSignIn();
                }
              }}
            >
              {Config.selfRegistration.needsAccountModalSubText}
            </Content>
          </div>
        </div>
      </ModalZeroButton>
    </>
  );
};
