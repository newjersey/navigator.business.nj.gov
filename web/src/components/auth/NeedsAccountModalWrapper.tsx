import { NeedsAccountContext } from "@/contexts/needsAccountContext";
import { IsAuthenticated } from "@/lib/auth/AuthContext";
import { useMountEffectWhenDefined } from "@/lib/utils/helpers";
import { ReactElement, ReactNode, useContext } from "react";

export const NeedsAccountModalWrapper = (props: { children: ReactNode }): ReactElement => {
  const { isAuthenticated, setShowNeedsAccountModal } = useContext(NeedsAccountContext);
  useMountEffectWhenDefined(() => {
    if (isAuthenticated !== IsAuthenticated.TRUE) {
      setShowNeedsAccountModal(true);
    }
  }, isAuthenticated);

  if (isAuthenticated !== IsAuthenticated.TRUE) {
    return (
      <div className="disabled-overlay">
        <div className="cursor-wrapper">{props.children}</div>
      </div>
    );
  }
  return <>{props.children}</>;
};
