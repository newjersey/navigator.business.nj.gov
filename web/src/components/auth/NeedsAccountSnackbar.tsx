import { Content } from "@/components/Content";
import { Heading } from "@/components/njwds-extended/Heading";
import { SnackbarAlert } from "@/components/njwds-extended/SnackbarAlert";
import { Icon } from "@/components/njwds/Icon";
import { AuthContext } from "@/contexts/authContext";
import { NeedsAccountContext } from "@/contexts/needsAccountContext";
import { useConfig } from "@/lib/data-hooks/useConfig";
import { useSidebarCards } from "@/lib/data-hooks/useSidebarCards";
import { MediaQueries } from "@/lib/PageSizes";
import { SIDEBAR_CARDS } from "@businessnjgovnavigator/shared/domain-logic/sidebarCards";
import { IconButton, useMediaQuery } from "@mui/material";
import { ReactElement, useContext } from "react";

export const NeedsAccountSnackbar = (): ReactElement => {
  const { Config } = useConfig();
  const { showCard } = useSidebarCards();
  const { showNeedsAccountSnackbar, setShowNeedsAccountSnackbar } = useContext(NeedsAccountContext);
  const { state } = useContext(AuthContext);
  const isDesktopAndUp = useMediaQuery(MediaQueries.desktopAndUp);

  if (!showNeedsAccountSnackbar) {
    return <></>;
  }

  const handleClose = async (): Promise<void> => {
    setShowNeedsAccountSnackbar(false);
    if (state.activeUser?.encounteredMyNjLinkingError) {
      await showCard(SIDEBAR_CARDS.notRegisteredExistingAccount);
    } else {
      await showCard(SIDEBAR_CARDS.notRegistered);
    }
  };

  const getTitle = (): string => {
    return state.activeUser?.encounteredMyNjLinkingError
      ? Config.navigationDefaults.needsAccountSnackbarTitleExistingAccount
      : Config.navigationDefaults.needsAccountSnackbarTitle;
  };

  const getBody = (): string => {
    return state.activeUser?.encounteredMyNjLinkingError
      ? Config.navigationDefaults.needsAccountSnackbarBodyExistingAccount
      : Config.navigationDefaults.needsAccountSnackbarBody;
  };

  return (
    <SnackbarAlert
      isOpen={showNeedsAccountSnackbar}
      close={(): void => setShowNeedsAccountSnackbar(false)}
      variant="info"
      noIcon={true}
      autoHideDuration={null}
      className={"bg-base-lightest"}
      snackBarProps={{ sx: { paddingX: 0 } }}
      dataTestid="self-reg-snackbar"
    >
      <div className="fin fac padding-y-2">
        {isDesktopAndUp ? (
          <img
            src={`/img/needs-account-snackbar-icon.svg`}
            alt="registration"
            style={{ marginRight: "20px", width: "51px", height: "64px" }}
          />
        ) : (
          <></>
        )}
        <div>
          <Heading level={3} className="padding-right-2">
            {getTitle()}
          </Heading>
          <IconButton
            aria-label="close"
            onClick={handleClose}
            sx={{
              position: "absolute",
              right: 0,
              top: 0,
              color: "#757575",
            }}
          >
            <Icon className="usa-icon--size-4">close</Icon>
          </IconButton>
          <Content>{getBody()}</Content>
        </div>
      </div>
    </SnackbarAlert>
  );
};
