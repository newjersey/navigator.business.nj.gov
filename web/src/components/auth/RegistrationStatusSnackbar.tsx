import { Content } from "@/components/Content";
import { AlertVariant } from "@/components/njwds-extended/Alert";
import { Heading } from "@/components/njwds-extended/Heading";
import { SnackbarAlert } from "@/components/njwds-extended/SnackbarAlert";
import { AuthContext } from "@/contexts/authContext";
import { NeedsAccountContext } from "@/contexts/needsAccountContext";
import { IsAuthenticated } from "@/lib/auth/AuthContext";
import { useConfig } from "@/lib/data-hooks/useConfig";
import analytics from "@/lib/utils/analytics";
import { RegistrationStatus } from "@businessnjgovnavigator/shared/";
import { ReactElement, useContext, useEffect } from "react";

export const RegistrationStatusSnackbar = (): ReactElement => {
  const { isAuthenticated, registrationStatus, setRegistrationStatus } =
    useContext(NeedsAccountContext);
  const { state } = useContext(AuthContext);
  const { Config } = useConfig();

  useEffect(() => {
    if (registrationStatus === "IN_PROGRESS" && isAuthenticated === IsAuthenticated.TRUE) {
      analytics.event.roadmap_dashboard.arrive.arrive_from_myNJ_registration();
      setRegistrationStatus("SUCCESS");
    }
  }, [isAuthenticated, registrationStatus, setRegistrationStatus]);

  type AlertStatus = Exclude<RegistrationStatus, "IN_PROGRESS">;
  const alertMap: Record<AlertStatus, AlertVariant> = {
    SUCCESS: "success",
    DUPLICATE_ERROR: "error",
    RESPONSE_ERROR: "error",
  };

  const contentMap: Record<AlertStatus, string> = {
    SUCCESS: Config.selfRegistration.accountSuccessSnackbarBody,
    DUPLICATE_ERROR: Config.selfRegistration.errorTextDuplicateSignUp,
    RESPONSE_ERROR: Config.selfRegistration.errorTextGeneric,
  };

  const getSuccessTitle = (): string => {
    return state.activeUser?.encounteredMyNjLinkingError
      ? Config.selfRegistration.accountSuccessSnackbarTitleExistingAccount
      : Config.selfRegistration.accountSuccessSnackbarTitle;
  };

  if (
    !registrationStatus ||
    registrationStatus === "IN_PROGRESS" ||
    state.activeUser?.id === state.activeUser?.email
  ) {
    return <></>;
  }

  if (alertMap[registrationStatus] === "success") {
    return (
      <SnackbarAlert
        isOpen={!!registrationStatus}
        close={(): void => setRegistrationStatus(undefined)}
        variant={alertMap[registrationStatus]}
        noIcon={true}
        autoHideDuration={7000}
        snackBarProps={{ sx: { paddingX: 0 } }}
        dataTestid="reg-snackbar"
      >
        <div className={"fin fac padding-y-2 padding-right-3"}>
          <img
            src={`/img/congratulations-green.svg`}
            alt="congratulations"
            data-testid={"congratulations-logo"}
            style={{ marginRight: "20px", width: "64px", height: "64px" }}
          />
          <div>
            <Heading level={3} className="padding-right-2">
              {getSuccessTitle()}
            </Heading>
            <Content>{contentMap[registrationStatus]}</Content>
          </div>
        </div>
      </SnackbarAlert>
    );
  }
  return (
    <SnackbarAlert
      isOpen={!!registrationStatus}
      close={(): void => setRegistrationStatus(undefined)}
      closeIcon={true}
      variant={alertMap[registrationStatus]}
      autoHideDuration={null}
      dataTestid="reg-snackbar"
      heading={Config.profileDefaults.default.errorTextHeader}
    >
      <Content>{contentMap[registrationStatus]}</Content>
    </SnackbarAlert>
  );
};
