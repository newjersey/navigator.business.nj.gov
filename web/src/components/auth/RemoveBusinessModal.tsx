import { useConfig } from "@/lib/data-hooks/useConfig";
import { templateEval, useMountEffectWhenDefined } from "@/lib/utils/helpers";
import { ReactElement, useContext, useRef, useState } from "react";
import { RemoveBusinessContext } from "@/contexts/removeBusinessContext";
import { Checkbox, FormControlLabel, useMediaQuery } from "@mui/material";
import { MediaQueries } from "@/lib/PageSizes";
import { ReverseOrderInMobile } from "@/components/njwds-layout/ReverseOrderInMobile";
import { SecondaryButton } from "@/components/njwds-extended/SecondaryButton";
import { PrimaryButton } from "@/components/njwds-extended/PrimaryButton";
import { ModalZeroButton } from "@/components/ModalZeroButton";
import { LiveChatHelpButton } from "@/components/njwds-extended/LiveChatHelpButton";
import { Content } from "@/components/Content";
import { WithErrorBar } from "@/components/WithErrorBar";
import { Alert } from "@/components/njwds-extended/Alert";
import { Business } from "@businessnjgovnavigator/shared/userData";
import { getNavBarBusinessTitle } from "@/lib/domain-logic/getNavBarBusinessTitle";
import { useUserData } from "@/lib/data-hooks/useUserData";
import { IsAuthenticated } from "@/lib/auth/AuthContext";
import { AuthContext } from "@/contexts/authContext";
import analytics from "@/lib/utils/analytics";

interface Props {
  CMS_ONLY_fakeBusiness?: Business;
}

export const RemoveBusinessModal = (props: Props): ReactElement => {
  const userDataFromHook = useUserData();
  const business = props.CMS_ONLY_fakeBusiness ?? userDataFromHook.business;
  const { state } = useContext(AuthContext);
  const { showRemoveBusinessModal, setShowRemoveBusinessModal } = useContext(RemoveBusinessContext);
  const { Config } = useConfig();

  const errorRef = useRef<HTMLDivElement>(null);

  const [checked, setChecked] = useState(false);
  const [hasRemoveBusinessError, setRemoveBusinessError] = useState(false);

  useMountEffectWhenDefined(() => {
    setShowRemoveBusinessModal(false);
  }, setShowRemoveBusinessModal);

  const isMobileAndUp = useMediaQuery(MediaQueries.mobileAndUp);

  const businessName: string = getNavBarBusinessTitle(
    business,
    state.isAuthenticated === IsAuthenticated.TRUE,
  );

  const close = (): void => {
    setChecked(false);
    setRemoveBusinessError(false);
    setShowRemoveBusinessModal(false);
  };

  const removeBusiness = (): void => {
    if (!checked) {
      setRemoveBusinessError(true);
      errorRef.current?.focus();
      return;
    }

    close();
  };
  const buttonNode = (
    <div
      className={`padding-x-4 padding-y-3 ${
        isMobileAndUp ? "flex flex-column flex-align-end" : ""
      }`}
      data-testid="modal-content"
    >
      <ReverseOrderInMobile>
        <>
          <LiveChatHelpButton
            analyticsEvent={analytics.event.remove_business_modal_help_button.click.open_live_chat}
          />
          <SecondaryButton
            isColor="border-dark-red"
            dataTestId="modal-button-secondary"
            onClick={close}
          >
            {Config.removeBusinessModal.cancelButtonText}
          </SecondaryButton>
          <div className="margin-bottom-2 mobile-lg:margin-bottom-0">
            <PrimaryButton
              isColor="error"
              isRightMarginRemoved={true}
              onClick={removeBusiness}
              dataTestId="modal-button-primary"
            >
              {Config.removeBusinessModal.removeBusinessButtonText}
            </PrimaryButton>
          </div>
        </>
      </ReverseOrderInMobile>
    </div>
  );

  return (
    <ModalZeroButton
      isOpen={showRemoveBusinessModal}
      close={close}
      title={Config.removeBusinessModal.header}
      unpaddedChildren={buttonNode}
    >
      {hasRemoveBusinessError && (
        <Alert variant="error" dataTestid={"profile-error-alert"} ref={errorRef}>
          <div>{Config.removeBusinessModal.agreementCheckboxErrorText}</div>
        </Alert>
      )}

      <Content>
        {templateEval(Config.removeBusinessModal.filingOfficialBusinessClosureText, {
          businessName: businessName,
        })}
      </Content>

      <div className={"padding-5px"}>
        <Content>
          {templateEval(Config.removeBusinessModal.businessRemovalText, {
            businessName: businessName,
          })}
        </Content>
      </div>

      <div className={"padding-5px"}>{Config.removeBusinessModal.irreversibleOperationText}</div>

      <div className={"padding-205"}>
        <WithErrorBar hasError={hasRemoveBusinessError} type={"ALWAYS"}>
          <FormControlLabel
            label={Config.removeBusinessModal.agreementCheckboxText}
            control={
              <Checkbox
                data-testid="agree-checkbox"
                checked={checked}
                onChange={(e) => {
                  setChecked(e.target.checked);
                  if (e.target.checked) setRemoveBusinessError(false);
                }}
                className={
                  hasRemoveBusinessError ? "agreement-checkbox-error" : "agreement-checkbox-base"
                }
              />
            }
          />
        </WithErrorBar>
      </div>
      <hr />
    </ModalZeroButton>
  );
};
