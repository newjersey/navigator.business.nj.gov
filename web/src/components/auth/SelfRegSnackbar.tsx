import { Content } from "@/components/Content";
import { AlertVariant } from "@/components/njwds-extended/Alert";
import { SnackbarAlert } from "@/components/njwds-extended/SnackbarAlert";
import { Icon } from "@/components/njwds/Icon";
import { AuthAlertContext } from "@/contexts/authAlertContext";
import { AuthContext } from "@/contexts/authContext";
import { IsAuthenticated } from "@/lib/auth/AuthContext";
import { useConfig } from "@/lib/data-hooks/useConfig";
import analytics from "@/lib/utils/analytics";
import { RegistrationStatus } from "@businessnjgovnavigator/shared/";
import { IconButton } from "@mui/material";
import { ReactElement, useContext, useEffect } from "react";

export const SelfRegSnackbar = (): ReactElement => {
  const { isAuthenticated, registrationAlertStatus, setRegistrationAlertStatus } =
    useContext(AuthAlertContext);
  const { state } = useContext(AuthContext);
  const { Config } = useConfig();

  useEffect(() => {
    if (registrationAlertStatus === "IN_PROGRESS" && isAuthenticated === IsAuthenticated.TRUE) {
      analytics.event.roadmap_dashboard.arrive.arrive_from_myNJ_registration();
      setRegistrationAlertStatus("SUCCESS");
    }
  }, [isAuthenticated, registrationAlertStatus, setRegistrationAlertStatus]);

  type AlertStatus = Exclude<RegistrationStatus, "IN_PROGRESS">;
  const alertMap: Record<AlertStatus, AlertVariant> = {
    SUCCESS: "success",
    DUPLICATE_ERROR: "error",
    RESPONSE_ERROR: "error",
  };

  const contentMap: Record<AlertStatus, string> = {
    SUCCESS: Config.navigationDefaults.guestSuccessBody,
    DUPLICATE_ERROR: Config.selfRegistration.errorTextDuplicateSignUp,
    RESPONSE_ERROR: Config.selfRegistration.errorTextGeneric,
  };

  const getSuccessTitle = (): string => {
    return state.activeUser?.encounteredMyNjLinkingError
      ? Config.navigationDefaults.guestSuccessTitleExistingAccount
      : Config.navigationDefaults.guestSuccessTitle;
  };

  if (!registrationAlertStatus || registrationAlertStatus === "IN_PROGRESS") {
    return <></>;
  }

  if (alertMap[registrationAlertStatus] === "success") {
    return (
      <SnackbarAlert
        isOpen={!!registrationAlertStatus}
        close={(): void => setRegistrationAlertStatus(undefined)}
        variant={alertMap[registrationAlertStatus]}
        noIcon={true}
        autoHideDuration={null}
        snackBarProps={{ sx: { paddingX: 0 } }}
        dataTestid="reg-snackbar"
      >
        <div className={"fin fac padding-y-2 padding-right-3"}>
          <img
            src={`/img/congratulations-green.svg`}
            alt="congratulations"
            data-testid={"congratulations-logo"}
            style={{ marginRight: "20px", width: "64px", height: "64px" }}
          />
          <div>
            <h3 className="padding-right-2">{getSuccessTitle()}</h3>
            <IconButton
              aria-label="close"
              onClick={(): void => setRegistrationAlertStatus(undefined)}
              sx={{
                position: "absolute",
                right: 0,
                top: 0,
                color: "#757575",
              }}
            >
              <Icon className={"usa-icon--size-4"}>close</Icon>
            </IconButton>
            <Content>{contentMap[registrationAlertStatus]}</Content>
          </div>
        </div>
      </SnackbarAlert>
    );
  }
  return (
    <SnackbarAlert
      isOpen={!!registrationAlertStatus}
      close={(): void => setRegistrationAlertStatus(undefined)}
      closeIcon={true}
      variant={alertMap[registrationAlertStatus]}
      autoHideDuration={null}
      dataTestid="reg-snackbar"
      heading={Config.profileDefaults.default.errorTextHeader}
    >
      <Content>{contentMap[registrationAlertStatus]}</Content>
    </SnackbarAlert>
  );
};
