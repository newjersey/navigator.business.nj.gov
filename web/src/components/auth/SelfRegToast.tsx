import { Content } from "@/components/Content";
import { AlertVariant } from "@/components/njwds-extended/Alert";
import { ToastAlert } from "@/components/njwds-extended/ToastAlert";
import { Icon } from "@/components/njwds/Icon";
import { IsAuthenticated } from "@/lib/auth/AuthContext";
import analytics from "@/lib/utils/analytics";
import { AuthAlertContext } from "@/pages/_app";
import Config from "@businessnjgovnavigator/content/fieldConfig/config.json";
import { RegistrationStatus } from "@businessnjgovnavigator/shared/";
import { IconButton } from "@mui/material";
import React, { ReactElement, useContext, useEffect } from "react";

export const SelfRegToast = (): ReactElement => {
  const { isAuthenticated, registrationAlertStatus, setRegistrationAlertStatus } =
    useContext(AuthAlertContext);

  useEffect(() => {
    if (registrationAlertStatus == "IN_PROGRESS" && isAuthenticated == IsAuthenticated.TRUE) {
      analytics.event.roadmap_dashboard.arrive.arrive_from_myNJ_registration();
      setRegistrationAlertStatus("SUCCESS");
    }
  }, [isAuthenticated, registrationAlertStatus, setRegistrationAlertStatus]);

  type AlertStatus = Exclude<RegistrationStatus, "IN_PROGRESS">;
  const alertMap: Record<AlertStatus, AlertVariant> = {
    SUCCESS: "success",
    DUPLICATE_ERROR: "error",
    RESPONSE_ERROR: "error",
  };

  const contentMap: Record<AlertStatus, string> = {
    SUCCESS: Config.navigationDefaults.guestSuccessBody,
    DUPLICATE_ERROR: Config.selfRegistration.errorTextDuplicateSignUp,
    RESPONSE_ERROR: Config.selfRegistration.errorTextGeneric,
  };

  if (!registrationAlertStatus || registrationAlertStatus === "IN_PROGRESS") return <></>;

  return (
    <ToastAlert
      isOpen={!!registrationAlertStatus}
      close={() => setRegistrationAlertStatus(undefined)}
      variant={alertMap[registrationAlertStatus]}
      noIcon={true}
      autoHideDuration={null}
      snackBarProps={{ sx: { paddingX: 0 } }}
    >
      <div className="fin fac padding-y-2 padding-right-3" data-testid={"reg-toast"}>
        {alertMap[registrationAlertStatus] == "success" ? (
          <img
            src={`/img/congratulations-green.svg`}
            alt="congratulations"
            data-testid={"congratulations-logo"}
            style={{ marginRight: "20px", width: "64px", height: "64px" }}
          />
        ) : (
          <></>
        )}
        <div>
          {alertMap[registrationAlertStatus] == "success" ? (
            <Content className="padding-right-2">{Config.navigationDefaults.guestSuccessTitle}</Content>
          ) : (
            <></>
          )}
          <IconButton
            aria-label="close"
            onClick={() => setRegistrationAlertStatus(undefined)}
            sx={{
              position: "absolute",
              right: 0,
              top: 0,
              color: (theme) => theme.palette.grey[500],
            }}
          >
            <Icon className="usa-icon--size-4">close</Icon>
          </IconButton>
          <Content>{contentMap[registrationAlertStatus]}</Content>
        </div>
      </div>
    </ToastAlert>
  );
};
