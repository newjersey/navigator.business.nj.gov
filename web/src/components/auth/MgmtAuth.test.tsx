import { MgmtAuth } from "@/components/auth/MgmtAuth";
import * as api from "@/lib/api-client/apiClient";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

jest.mock("@/lib/api-client/apiClient", () => ({ post: jest.fn() }));

const mockApi = api as jest.Mocked<typeof api>;

describe("<MgmtAuth />", () => {
  beforeEach(() => {
    jest.resetAllMocks();
    password = "";
    isAuthed = false;
  });

  let password = "";
  const setPassword = (value: string): void => {
    password = value;
  };
  let isAuthed = false;
  const setIsAuthed = (value: boolean): void => {
    isAuthed = value;
  };

  it("sets the password field on change", async () => {
    mockApi.post.mockReturnValue(Promise.resolve());
    const user = userEvent.setup();
    render(<MgmtAuth password={password} setPassword={setPassword} setIsAuthed={setIsAuthed} />);
    const pwField = screen.getByTestId("mgmt-password-field");
    await user.type(pwField, "1234");
    expect(password).toBe("1234");
  });

  it("calls the api on submission and sets isAuthed to true on success", async () => {
    mockApi.post.mockReturnValue(Promise.resolve());
    const user = userEvent.setup();
    render(<MgmtAuth password={password} setPassword={setPassword} setIsAuthed={setIsAuthed} />);
    const pwField = screen.getByTestId("mgmt-password-field");
    await user.type(pwField, "1234");
    await user.click(screen.getByTestId("mgmt-submit-bttn"));
    await waitFor(() => {
      return expect(mockApi.post).toHaveBeenCalled();
    });
    expect(isAuthed).toEqual(true);
  });

  it("submits the form when clicking enter", async () => {
    mockApi.post.mockReturnValue(Promise.resolve());
    const user = userEvent.setup();
    render(<MgmtAuth password={password} setPassword={setPassword} setIsAuthed={setIsAuthed} />);
    const pwField = screen.getByTestId("mgmt-password-field");
    await user.type(pwField, "1234");
    await user.keyboard("{enter}");
    await waitFor(() => {
      return expect(mockApi.post).toHaveBeenCalled();
    });
    expect(isAuthed).toEqual(true);
  });

  it("calls the api on submission and doesn't change isAuthed to true on failure", async () => {
    mockApi.post.mockReturnValue(Promise.reject());

    render(<MgmtAuth password={password} setPassword={setPassword} setIsAuthed={setIsAuthed} />);
    const pwField = screen.getByTestId("mgmt-password-field");
    fireEvent.change(pwField, { target: { value: "1234" } });
    fireEvent.click(screen.getByTestId("mgmt-submit-bttn"));
    await waitFor(() => {
      return expect(screen.getByText("Authentication failed")).toBeInTheDocument();
    });
    expect(mockApi.post).toHaveBeenCalled();
    expect(isAuthed).toEqual(false);
  });
});
