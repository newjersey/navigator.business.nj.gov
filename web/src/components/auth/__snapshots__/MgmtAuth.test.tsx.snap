// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MgmtAuth /> matches the snapshot 1`] = `
<body>
  <div>
    <h1>
      Enter admin password:
    </h1>
    <label
      for="password"
    >
      Password
    </label>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <div
        class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          data-testid="mgmt-password-field"
          id="password"
          name="password"
          type="password"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-hdw1oc"
          >
            <span
              class="notranslate"
            >
              â€‹
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <button
      class="usa-button margin-top-2"
      data-testid="mgmt-submit-bttn"
    >
      Submit
    </button>
  </div>
</body>
`;
